packages/flutter_tools/lib/src/custom_devices/custom_device.dart

```diff
@@ -0,0 +1,725 @@
+      platform: await targetPlatform,
+      buildInfo: debuggingOptions.buildInfo,
+      mainPath: mainPath,
+      depfilePath: defaultDepfilePath,
+      assetDirPath: assetBundleDir,
+      treeShakeIcons: false,
+    );
+
+    // if we have a post build step (needed for some embedders), execute it
+    if (_config.postBuildCommand != null) {
+      await _tryPostBuild(
+        appName: package.name,
+        localPath: assetBundleDir,
+      );
+    }
+
+    // install the app on the device
+    // (will invoke the uninstall and then the install command internally)
+    await installApp(package, userIdentifier: userIdentifier);
+
+    // finally launch the app
+    return _getOrCreateAppSession(package).start(
+      mainPath: mainPath,
+      route: route,
+      debuggingOptions: debuggingOptions,
+      platformArgs: platformArgs,
+      prebuiltApplication: prebuiltApplication,
+      ipv6: ipv6,
+      userIdentifier: userIdentifier,
+    );
+  }
+
+  @override
+  Future<bool> stopApp(covariant ApplicationPackage app, {String userIdentifier}) {
+    return _getOrCreateAppSession(app).stop();
+  }
+
+  @override
+  // TODO(ardera): Allow configuring or auto-detecting the target platform, https://github.com/flutter/flutter/issues/78151
+  Future<TargetPlatform> get targetPlatform async => TargetPlatform.linux_arm64;
+
+  @override
+  Future<bool> uninstallApp(covariant ApplicationPackage app, {String userIdentifier}) {
+    return _tryUninstall(appName: app.name);
+  }
+}
+
+class CustomDevices extends PollingDeviceDiscovery {
+  /// Create a custom device discovery that using the [CustomDeviceConfig]
+  /// in the context.
+  CustomDevices({
+    @required FeatureFlags featureFlags,
+    @required ProcessManager processManager,
+    @required Logger logger,
+  }) : _customDeviceWorkflow = CustomDeviceWorkflow(
+        featureFlags: featureFlags,
+      ),
+       _logger = logger,
+       _processManager = processManager,
+       _config = null,
+       super('custom devices');
+
+  /// Create a custom device discovery for the given [CustomDevicesConfig] instead
+  /// of sourcing it from the context.
+  @visibleForTesting
+  CustomDevices.test({
+    @required FeatureFlags featureFlags,
+    @required ProcessManager processManager,
+    @required Logger logger,
+    @required CustomDevicesConfig config
+  }) : _customDeviceWorkflow = CustomDeviceWorkflow(
+        featureFlags: featureFlags,
+      ),
+      _logger = logger,
+      _processManager = processManager,
+      _config = config,
+      super('custom devices');
+
+  final CustomDeviceWorkflow  _customDeviceWorkflow;
+  final ProcessManager _processManager;
+  final Logger _logger;
+  final CustomDevicesConfig _config;
+
+  @override
+  bool get supportsPlatform => true;
+
+  @override
+  bool get canListAnything => _customDeviceWorkflow.canListDevices;
+
+  CustomDevicesConfig get customDevicesConfig => _config ?? context.get<CustomDevicesConfig>();
```

Comment:
Avoid introducing `context` getters/setters. Instead ensure config is always provided

Reply:
Okay, yeah config is now not available in the context anymore and instead constructed when the `FlutterDeviceManager` constructer is called in `context_runner.dart:199`